{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### \n",
    "\n",
    "In this nb, we create a class to contain GO's go term tree (the term ontology)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os.path\n",
    "import wget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "obo_url = 'http://current.geneontology.org/ontology/go-basic.obo'\n",
    "obo_fp = 'data/go-basic.obo'\n",
    "if not os.path.exists(obo_fp):\n",
    "    wget.download(obo_url, obo_fp)\n",
    "    print('downloaded %s' % obo_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "format-version: 1.2\n",
      "data-version: releases/2020-02-21\n",
      "subsetdef: gocheck_do_not_annotate \"Term not to be used for direct annotation\"\n",
      "subsetdef: gocheck_do_not_manually_annotate \"Term not to be used for direct manual annotation\"\n",
      "subsetdef: goslim_agr \"AGR slim\"\n",
      "subsetdef: goslim_aspergillus \"Aspergillus GO slim\"\n",
      "subsetdef: goslim_candida \"Candida GO slim\"\n",
      "subsetdef: goslim_chembl \"ChEMBL protein targets summary\"\n",
      "subsetdef: goslim_drosophila \"Drosophila GO slim\"\n",
      "subsetdef: goslim_flybase_ribbon \"FlyBase Drosophila GO ribbon slim\"\n",
      "subsetdef: goslim_generic \"Generic GO slim\"\n",
      "subsetdef: goslim_metagenomics \"Metagenomics GO slim\"\n",
      "subsetdef: goslim_mouse \"Mouse GO slim\"\n",
      "subsetdef: goslim_pir \"PIR GO slim\"\n",
      "subsetdef: goslim_plant \"Plant GO slim\"\n",
      "subsetdef: goslim_pombe \"Fission yeast GO slim\"\n",
      "subsetdef: goslim_synapse \"synapse GO slim\"\n",
      "subsetdef: goslim_yeast \"Yeast GO slim\"\n",
      "synonymtypedef: syngo_official_label \"label approved by the SynGO project\"\n",
      "synonymtypedef: systematic_synonym \"Systematic synonym\" EXACT\n",
      "default-namespace: gene_ontology\n",
      "remark: cvs version: use data-version\n",
      "remark: Includes Ontology(OntologyID(OntologyIRI(<http://purl.obolibrary.org/obo/go/never_in_taxon.owl>))) [Axioms: 18 Logical Axioms: 0]\n",
      "ontology: go\n",
      "\n",
      "[Term]\n",
      "id: GO:0000001\n",
      "name: mitochondrion inheritance\n",
      "namespace: biological_process\n",
      "def: \"The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:11389764]\n",
      "synonym: \"mitochondrial inheritance\" EXACT []\n",
      "is_a: GO:0048308 ! organelle inheritance\n",
      "is_a: GO:0048311 ! mitochondrion distribution\n",
      "\n",
      "[Term]\n",
      "id: GO:0000002\n",
      "name: mitochondrial genome maintenance\n",
      "namespace: biological_process\n",
      "def: \"The maintenance of the structure and integrity of the mitochondrial genome; includes replication and segregation of the mitochondrial chromosome.\" [GOC:ai, GOC:vw]\n",
      "is_a: GO:0007005 ! mitochondrion organization\n",
      "\n",
      "[Term]\n",
      "id: GO:0000003\n",
      "name: reproduction\n",
      "namespace: biological_process\n",
      "alt_id: GO:0019952\n",
      "alt_id: GO:0050876\n",
      "def: \"The production of new individuals that contain some portion of genetic material inherited from one or more parent organisms.\" [GOC:go_curators, GOC:isa_complete, GOC:jl, ISBN:0198506732]\n",
      "subset: goslim_agr\n",
      "subset: goslim_chembl\n",
      "subset: goslim_flybase_ribbon\n",
      "subset: goslim_generic\n",
      "subset: goslim_pir\n",
      "subset: goslim_plant\n",
      "synonym: \"reproductive physiological process\" EXACT []\n",
      "xref: Wikipedia:Reproduction\n",
      "is_a: GO:0008150 ! biological_process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000005\n",
      "name: obsolete ribosomal chaperone activity\n",
      "namespace: molecular_function\n",
      "def: \"OBSOLETE. Assists in the correct assembly of ribosomes or ribosomal subunits in vivo, but is not a component of the assembled ribosome when performing its normal biological function.\" [GOC:jl, PMID:12150913]\n",
      "comment: This term was made obsolete because it refers to a class of gene products and a biological process rather than a molecular function.\n",
      "synonym: \"ribosomal chaperone activity\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0042254\n",
      "consider: GO:0044183\n",
      "consider: GO:0051082\n",
      "\n",
      "[Term]\n",
      "id: GO:0000006\n",
      "name: high-affinity zinc transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "def: \"Enables the transfer of zinc ions (Zn2+) from one side of a membrane to the other, probably powered by proton motive force. In high-affinity transport the transporter is able to bind the solute even if it is only present at very low concentrations.\" [TC:2.A.5.1.1]\n",
      "synonym: \"high affinity zinc uptake transmembrane transporter activity\" EXACT []\n",
      "synonym: \"high-affinity zinc uptake transmembrane transporter activity\" RELATED []\n",
      "is_a: GO:0005385 ! zinc ion transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000007\n",
      "name: low-affinity zinc ion transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "def: \"Enables the transfer of a solute or solutes from one side of a membrane to the other according to the reaction: Zn2+ = Zn2+, probably powered by proton motive force. In low-affinity transport the transporter is able to bind the solute only if it is present at very high concentrations.\" [GOC:mtg_transport, ISBN:0815340729]\n",
      "is_a: GO:0005385 ! zinc ion transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000008\n",
      "name: obsolete thioredoxin\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0000013\n",
      "def: \"OBSOLETE. A small disulfide-containing redox protein that serves as a general protein disulfide oxidoreductase. Interacts with a broad range of proteins by a redox mechanism, based on the reversible oxidation of 2 cysteine thiol groups to a disulfide, accompanied by the transfer of 2 electrons and 2 protons. The net result is the covalent interconversion of a disulfide and a dithiol.\" [GOC:kd]\n",
      "comment: This term was made obsolete because it represents gene products.\n",
      "synonym: \"thioredoxin\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0003756\n",
      "consider: GO:0015036\n",
      "\n",
      "[Term]\n",
      "id: GO:0000009\n",
      "name: alpha-1,6-mannosyltransferase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the transfer of a mannose residue to an oligosaccharide, forming an alpha-(1->6) linkage.\" [GOC:mcc, PMID:2644248]\n",
      "synonym: \"1,6-alpha-mannosyltransferase activity\" EXACT []\n",
      "xref: EC:2.4.1.232\n",
      "xref: Reactome:R-HSA-449718 \"Addition of a third mannose to the N-glycan precursor by ALG2\"\n",
      "is_a: GO:0000030 ! mannosyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000010\n",
      "name: trans-hexaprenyltranstransferase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the reaction: all-trans-hexaprenyl diphosphate + isopentenyl diphosphate = all-trans-heptaprenyl diphosphate + diphosphate.\" [PMID:9708911]\n",
      "xref: KEGG_REACTION:R05612\n",
      "xref: RHEA:20836\n",
      "is_a: GO:0016765 ! transferase activity, transferring alkyl or aryl (other than methyl) groups\n",
      "\n",
      "[Term]\n",
      "id: GO:0000011\n",
      "name: vacuole inheritance\n",
      "namespace: biological_process\n",
      "def: \"The distribution of vacuoles into daughter cells after mitosis or meiosis, mediated by interactions between vacuoles and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:14616069]\n",
      "is_a: GO:0007033 ! vacuole organization\n",
      "is_a: GO:0048308 ! organelle inheritance\n",
      "\n",
      "[Term]\n",
      "id: GO:0000012\n",
      "name: single strand break repair\n",
      "namespace: biological_process\n",
      "def: \"The repair of single strand breaks in DNA. Repair of such breaks is mediated by the same enzyme systems as are used in base excision repair.\" [http://www.ultranet.com/~jkimball/BiologyPages/D/DNArepair.html]\n",
      "is_a: GO:0006281 ! DNA repair\n",
      "\n",
      "[Term]\n",
      "id: GO:0000014\n",
      "name: single-stranded DNA endodeoxyribonuclease activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the hydrolysis of ester linkages within a single-stranded deoxyribonucleic acid molecule by creating internal breaks.\" [GOC:mah]\n",
      "synonym: \"single-stranded DNA specific endodeoxyribonuclease activity\" RELATED []\n",
      "synonym: \"ssDNA-specific endodeoxyribonuclease activity\" RELATED [GOC:mah]\n",
      "is_a: GO:0004520 ! endodeoxyribonuclease activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000015\n",
      "name: phosphopyruvate hydratase complex\n",
      "namespace: cellular_component\n",
      "def: \"A multimeric enzyme complex, usually a dimer or an octamer, that catalyzes the conversion of 2-phospho-D-glycerate to phosphoenolpyruvate and water.\" [GOC:jl, ISBN:0198506732]\n",
      "subset: goslim_metagenomics\n",
      "synonym: \"enolase complex\" EXACT []\n",
      "is_a: GO:1902494 ! catalytic complex\n",
      "relationship: part_of GO:0005829 ! cytosol\n",
      "\n",
      "[Term]\n",
      "id: GO:0000016\n",
      "name: lactase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the reaction: lactose + H2O = D-glucose + D-galactose.\" [PMID:12023280]\n",
      "synonym: \"lactase-phlorizin hydrolase activity\" BROAD []\n",
      "synonym: \"lactose galactohydrolase activity\" EXACT []\n",
      "xref: EC:3.2.1.108\n",
      "xref: MetaCyc:LACTASE-RXN\n",
      "xref: Reactome:R-HSA-189062 \"lactose + H2O => D-glucose + D-galactose\"\n",
      "xref: Reactome:R-HSA-5658001 \"Defective LCT does not hydrolyze Lac\"\n",
      "xref: RHEA:10076\n",
      "is_a: GO:0004553 ! hydrolase activity, hydrolyzing O-glycosyl compounds\n",
      "\n",
      "[Term]\n",
      "id: GO:0000017\n",
      "name: alpha-glucoside transport\n",
      "namespace: biological_process\n",
      "def: \"The directed movement of alpha-glucosides into, out of or within a cell, or between cells, by means of some agent such as a transporter or pore. Alpha-glucosides are glycosides in which the sugar group is a glucose residue, and the anomeric carbon of the bond is in an alpha configuration.\" [GOC:jl, http://www.biochem.purdue.edu/, ISBN:0198506732]\n",
      "is_a: GO:0042946 ! glucoside transport\n",
      "\n",
      "[Term]\n",
      "id: GO:0000018\n",
      "name: regulation of DNA recombination\n",
      "namespace: biological_process\n",
      "def: \"Any process that modulates the frequency, rate or extent of DNA recombination, a DNA metabolic process in which a new genotype is formed by reassortment of genes resulting in gene combinations different from those that were present in the parents.\" [GOC:go_curators, ISBN:0198506732]\n",
      "is_a: GO:0051052 ! regulation of DNA metabolic process\n",
      "relationship: regulates GO:0006310 ! DNA recombination\n",
      "\n",
      "[Term]\n",
      "id: GO:0000019\n",
      "name: regulation of mitotic recombination\n",
      "namespace: biological_process\n",
      "def: \"Any process that modulates the frequency, rate or extent of DNA recombination during mitosis.\" [GOC:go_curators]\n",
      "synonym: \"regulation of recombination within rDNA repeats\" NARROW []\n",
      "is_a: GO:0000018 ! regulation of DNA recombination\n",
      "relationship: regulates GO:0006312 ! mitotic recombination\n",
      "\n",
      "[Term]\n",
      "id: GO:0000020\n",
      "name: obsolete negative regulation of recombination within rDNA repeats\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. Any process that stops, prevents, or reduces the frequency, rate or extent of genetic recombination within the DNA of the genes coding for ribosomal RNA.\" [GOC:go_curators, ISBN:0198506732]\n",
      "comment: This term was made obsolete because it describes a substrate-specific process.\n",
      "synonym: \"negative regulation of recombination within rDNA repeats\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0045950\n",
      "\n",
      "[Term]\n",
      "200\n"
     ]
    }
   ],
   "source": [
    "#%%writefile example.txt\n",
    "\n",
    "obo_file = open(obo_fp, 'r')\n",
    "c=0\n",
    "for line in obo_file:\n",
    "    print(line.rstrip())\n",
    "    c+=1\n",
    "    if c==200:\n",
    "        break\n",
    "obo_file.close()\n",
    "print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
